psutil/arch/linux/users.c

```diff
@@ -5,13 +5,151 @@
  */
 
 #include <Python.h>
+#include <dlfcn.h>
 #include <utmp.h>
 
 #include "../../_psutil_common.h"
 
+/* Systemd function signatures that will be loaded */
+int (*sd_booted)(void);
+int (*sd_get_sessions)(char ***);
+int (*sd_session_get_leader)(const char *, pid_t *);
+int (*sd_session_get_remote_host)(const char *,char **);
+int (*sd_session_get_start_time)(const char *, uint64_t *);
+int (*sd_session_get_tty)(const char *, char **);
+int (*sd_session_get_username)(const char *, char **);
+
+#define dlsym_check(__h, __fn) do {           \
+    __fn = dlsym(__h, #__fn);                 \
+    if (dlerror() != NULL || __fn == NULL) {  \
+        dlclose(__h);                         \
+        return NULL;                          \
+    }                                         \
+} while (0)
+
+static void *
+load_systemd() {
+    void *handle = dlopen("libsystemd.so.0", RTLD_LAZY);
+    if (dlerror() != NULL || handle == NULL)
+        return NULL;
+
+    dlsym_check(handle, sd_booted);
+    dlsym_check(handle, sd_get_sessions);
+    dlsym_check(handle, sd_session_get_leader);
+    dlsym_check(handle, sd_session_get_remote_host);
+    dlsym_check(handle, sd_session_get_start_time);
+    dlsym_check(handle, sd_session_get_tty);
+    dlsym_check(handle, sd_session_get_username);
+
+    if (! sd_booted()) {
+        dlclose(handle);
```

Comment:
Please add  `psutil_debug("systemd not booted");` before returning.

Reply:
Done